services/spar/src/Spar/Sem/SAML2/SAML2WebSso.hs

```diff
@@ -0,0 +1,136 @@
+
+instance Member (Embed IO) r => MonadIO (Blah r) where
+  liftIO = Blah . embed @IO
+
+instance Member (Embed IO) r => HasCreateUUID (Blah r)
+
+instance Member (Embed IO) r => HasNow (Blah r)
+
+instance Members '[Error SparError, Final IO, AReqIDStore] r => SPStoreID AuthnRequest (Blah r) where
+  storeID = (wrapMonadClientBlah .) . AReqIDStore.store
+  unStoreID = wrapMonadClientBlah . AReqIDStore.unStore
+  isAliveID = wrapMonadClientBlah . AReqIDStore.isAlive
+
+instance Members '[Error SparError, Final IO, AssIDStore] r => SPStoreID Assertion (Blah r) where
+  storeID = (wrapMonadClientBlah .) . AssIDStore.store
+  unStoreID = wrapMonadClientBlah . AssIDStore.unStore
+  isAliveID = wrapMonadClientBlah . AssIDStore.isAlive
+
+instance Members '[Error SparError, IdPEffect.IdP, Final IO] r => SPStoreIdP SparError (Blah r) where
+  type IdPConfigExtra (Blah r) = WireIdP
+  type IdPConfigSPId (Blah r) = TeamId
+
+  storeIdPConfig = Blah . App.runSparInSem . App.storeIdPConfig
+  getIdPConfig = Blah . App.runSparInSem . App.getIdPConfig
+  getIdPConfigByIssuerOptionalSPId a = Blah . App.runSparInSem . App.getIdPConfigByIssuerOptionalSPId a
+
+instance Member (Error SparError) r => MonadError SparError (Blah r) where
+  throwError = Blah . throw
+  catchError m handler = Blah $ catch (unBlah m) $ unBlah . handler
+
+saml2ToSaml2WebSso ::
+  forall r a.
+  Members
+    '[ AReqIDStore,
+       AssIDStore,
+       Error SparError,
+       IdPEffect.IdP,
+       Input Opts,
+       Logger String,
+       Embed IO,
+       Final IO
+     ]
+    r =>
+  Sem (SAML2 ': r) a ->
+  Sem r a
+saml2ToSaml2WebSso =
+  interpretH $ \case
+    AuthReq n ma i -> do
+      get_a <- runT ma
+      ins <- getInspectorT
+      x <- raise $ unBlah $ SAML.authreq @_ @SparError n (inspectOrBomb ins get_a) i
+      s <- getInitialStateT
+      pure $ x <$ s
+    AuthResp mitlt ma mb mc ab -> do
+      get_a <- runT ma
+      get_b <- runT mb
+      get_c <- bindT $ uncurry mc
+      ins <- getInspectorT
+      s <- getInitialStateT
+
+      x <- raise $ unBlah $ SAML.authresp mitlt (inspectOrBomb ins get_a) (inspectOrBomb ins get_b) (\x y -> inspectOrBomb ins $ get_c $ (x, y) <$ s) ab
+      pure $ x <$ s
+    Meta t ma mb -> do
+      get_a <- runT ma
+      get_b <- runT mb
+      ins <- getInspectorT
+      x <- raise $ unBlah $ SAML.meta t (inspectOrBomb ins get_a) (inspectOrBomb ins get_b)
+      s <- getInitialStateT
+      pure $ x <$ s
+    ToggleCookie sbs mp -> do
+      liftT $ unBlah $ SAML.toggleCookie sbs mp
+
+inspectOrBomb ::
+  Members
+    '[ AReqIDStore,
+       AssIDStore,
+       Error SparError,
+       IdPEffect.IdP,
+       Logger String,
+       Input Opts,
+       Embed IO,
+       Final IO
+     ]
+    r =>
+  Inspector f ->
+  Sem (SAML2 : r) (f b) ->
+  Blah r b
+inspectOrBomb ins get_a = do
+  fa <- Blah $ saml2ToSaml2WebSso get_a
+  maybe
+    (error "saml2ToSaml2WebSso called with an uninspectable weaving functor")
```

Comment:
Scary! This case is impossible under normal circumstances, but can be triggered if any of the monadic arguments to the `SAML2` actions fail via `Polysemy.Error.throw`. A better solution here would be to instead throw an error of our own, but I wasn't sure which constructor to use --- due to a technical limitation in Polysemy, we can't rethrow (but the original error will arrive upstream anyway!)

What do you think is the right play here, @fisx?

Reply:
It's worth noting that this is entirely an artifact of the typeclass approach to `saml2-web-sso`, and will go away when we polysemize that library.

Reply:
how many places are there with monadic arguments?  can we refactor those away somehow?  or would that be harder than polysemizing saml2-web-sso?

Reply:
The issue are these two effects:

```haskell
data SAML2 m a where
  AuthReq ::
    NominalDiffTime ->
    m Issuer ->
    IdPId ->
    SAML2 m (FormRedirect AuthnRequest)
  AuthResp ::
    Maybe TeamId ->
    m Issuer ->
    m URI ->
    (AuthnResponse -> AccessVerdict -> m resp) ->
    AuthnResponseBody ->
    SAML2 m resp
```

their `m Blah` arguments, in particular.

The tactics effect in Polysemy lets us lift functions `a -> m b` into functions `f a -> m (f b)`, where `f` holds all of the accumulated state from other effects. In a world where `saml2-web-sso` were more flexible with the types of `authreq` and `authresp`, we could just push that `f` all the way through.

The behavior here is not surprising though. Imagine in everyday MTL land, where we pass `throwError blah` as the argument to `authreq`. That too would fail. The types in tactics just make it very clear (and, simultaneously, very obscure) to see where we have these sorts of data dependencies.

IMO, the solution here is to make this a `throw` instead of `error`. But I don't expect to see it, since `authreq` and `authresp` are only ever called called with `SparRoute` actions. 

Life (and this interpreter) would be much easier if these were just pure values instead of monadic ones, which we could do without polysemizing `saml2-web-sso` --- though maybe some callers depend on this being monadic?